<!-- Simulation of Statemachine with GMapping and Frontier Exploration -->

<launch>
	<arg name="robot_frame" default="robot0" />
	<arg name="autonomy_cmd_vel_topic"
		default="/robot0/autonomy/cmd_vel" />
	<arg name="teleoperation_cmd_vel_topic"
		default="/robot0/teleoperation/cmd_vel" />
	<arg name="cmd_vel_topic" default="/robot0/cmd_vel" />
	<arg name="goal_plugin" default="rsm::CalculateGoalState" />
	<arg name="navigation_plugin"
		default="rsm::NavigationState" />
	<arg name="mapping_plugin" default="rsm::MappingDummyState" />
	<arg name="waypoint_routines" default="['Reversing']" />
	<arg name="rviz" default="false" />

	<include file="$(find stdr_robot)/launch/robot_manager.launch" />

	<node type="stdr_server_node" pkg="stdr_server" name="stdr_server"
		output="screen"
		args="$(find stdr_resources)/maps/sparse_obstacles.yaml" />

	<node pkg="tf" type="static_transform_publisher" name="world2map"
		args="0 0 0 0 0 0  world map 100" />

	<include file="$(find stdr_gui)/launch/stdr_gui.launch" />

	<node pkg="stdr_robot" type="robot_handler"
		name="$(anon robot_spawn)"
		args="add $(find rsm_additions)/config/stdr_sim/robot1.yaml 1 2 0" />

	<!--include file="$(find stdr_launchers)/launch/server_with_map_and_gui_plus_robot.launch" 
		/ -->

	<include file="$(find rsm_core)/launch/rsm.launch">
		<arg name="update_frequency" value="20" />
		<arg name="robot_frame" value="$(arg robot_frame)" />
		<arg name="autonomy_cmd_vel_topic"
			value="$(arg autonomy_cmd_vel_topic)" />
		<arg name="teleoperation_cmd_vel_topic"
			value="$(arg teleoperation_cmd_vel_topic)" />
		<arg name="cmd_vel_topic" value="$(arg cmd_vel_topic)" />
		<arg name="calculate_goal_plugin" value="$(arg goal_plugin)" />
		<arg name="navigation_plugin" value="$(arg navigation_plugin)" />
		<arg name="mapping_plugin" value="$(arg mapping_plugin)" />
		<arg name="waypoint_routines" value="$(arg waypoint_routines)" />
		<arg name="exploration_goal_tolerance" value="0.05" />
	</include>

	<node pkg="rsm_additions" type="bootUpNode"
		name="bootUpNode" output="screen">
		<param name="wait_time" value="2" />
	</node>

	<node pkg="rsm_additions"
		type="additionsServiceProviderNode"
		name="additionsServiceProviderNode">
		<param name="update_frequency" value="20" />
		<param name="autonomy_cmd_vel_topic"
			value="$(arg autonomy_cmd_vel_topic)" />
		<param name="navigation_plugin" value="$(arg navigation_plugin)" />

	</node>

	<node pkg="gmapping" type="slam_gmapping" name="slam_gmapping">
		<param name="base_frame" value="$(arg robot_frame)" />
		<remap from="/scan" to="/robot0/laser_0" />
		<remap from="/map" to="/gmapping/map" />
	</node>

	<node pkg="move_base" type="move_base" respawn="false"
		name="move_base" output="screen">
		<rosparam
			file="$(find rsm_additions)/config/stdr_sim/costmap_common.yaml"
			command="load" ns="global_costmap" />
		<rosparam
			file="$(find rsm_additions)/config/stdr_sim/costmap_common.yaml"
			command="load" ns="local_costmap" />
		<rosparam
			file="$(find rsm_additions)/config/stdr_sim/costmap_local.yaml"
			command="load" />
		<rosparam
			file="$(find rsm_additions)/config/stdr_sim/costmap_global.yaml"
			command="load" />
		<rosparam
			file="$(find rsm_additions)/config/stdr_sim/planner.yaml"
			command="load" />
		<param name="global_costmap/robot_base_frame"
			value="$(arg robot_frame)" />
		<param name="local_costmap/robot_base_frame"
			value="$(arg robot_frame)" />
		<remap from="/map" to="/gmapping/map" />
		<remap from="/cmd_vel" to="$(arg autonomy_cmd_vel_topic)" />
	</node>

	<node pkg="move_base" type="move_base" respawn="false"
		name="move_base_reverse" output="screen">
		<rosparam
			file="$(find rsm_additions)/config/stdr_sim/costmap_common.yaml"
			command="load" ns="global_costmap" />
		<rosparam
			file="$(find rsm_additions)/config/stdr_sim/costmap_common.yaml"
			command="load" ns="local_costmap" />
		<rosparam
			file="$(find rsm_additions)/config/stdr_sim/costmap_local.yaml"
			command="load" />
		<rosparam
			file="$(find rsm_additions)/config/stdr_sim/costmap_global.yaml"
			command="load" />
		<rosparam
			file="$(find rsm_additions)/config/stdr_sim/planner.yaml"
			command="load" />
		<param name="global_costmap/robot_base_frame"
			value="$(arg robot_frame)_reverse" />
		<param name="local_costmap/robot_base_frame"
			value="$(arg robot_frame)_reverse" />
		<remap from="/map" to="/gmapping/map" />
		<remap from="/cmd_vel"
			to="$(arg autonomy_cmd_vel_topic)_reverse" />
		<remap from="move_base/goal" to="move_base_reverse/goal" />
		<remap from="move_base/cancel" to="move_base_reverse/cancel" />
		<remap from="move_base/feedback" to="move_base_reverse/feedback" />
		<remap from="move_base/status" to="move_base_reverse/status" />
		<remap from="move_base/result" to="move_base_reverse/result" />
	</node>

	<node pkg="tf" type="static_transform_publisher"
		name="robot0_reverse"
		args="0 0 0 3.1415 0 0             $(arg robot_frame)          $(arg robot_frame)_reverse		10" />

	<!-- Explore Lite -->
	<node pkg="explore_lite" type="explore" respawn="false"
		name="explore">
		<param name="robot_base_frame" value="$(arg robot_frame)" />
		<param name="costmap_topic" value="gmapping/map" />
		<param name="costmap_updates_topic" value="gmapping/map_updates" />
		<param name="visualize" value="true" />
		<param name="planner_frequency" value="5" />
		<param name="progress_timeout" value="3600.0" />
		<param name="potential_scale" value="3.0" />
		<param name="orientation_scale" value="0.0" />
		<param name="gain_scale" value="1.0" />
		<param name="transform_tolerance" value="0.3" />
		<param name="min_frontier_size" value="0.75" />
		<remap from="/move_base" to="/frontier_move_base" />
	</node>

	<node if="$(arg rviz)" pkg="rviz" type="rviz" name="rviz"
		args="-d $(find rsm_additions)/config/stdr_sim/rsm_simulation.rviz">
		<remap from="/move_base_simple/goal"
			to="/rsm/simpleGoal" />
	</node>
	<include if="$(arg rviz)"
		file="$(find rsm_rviz_plugins)/launch/rsm_rviz_plugins.launch">
		<arg name="update_frequency" value="20" />
	</include>
</launch>
